From ce36ab8d6955cc72dd7118a415be0e3cc444a8dc Mon Sep 17 00:00:00 2001
From: Murali Shankar <mshankar@slac.stanford.edu>
Date: Fri, 20 Feb 2015 15:22:15 -0800
Subject: [PATCH] If the button is down and we use the middle-click to drag the
 PV name, the toggle does not seem to reset. Ignore the drag event in case the
 button is down

---
 baselib/button.cc | 14 ++++++++++++++
 baselib/button.h  |  1 +
 2 files changed, 15 insertions(+)

diff --git a/baselib/button.cc b/baselib/button.cc
index f9710a0..4a5c16a 100644
--- a/baselib/button.cc
+++ b/baselib/button.cc
@@ -495,6 +495,7 @@ activeButtonClass::activeButtonClass ( void ) {
   strcpy( maxVisString, "" );
   connection.setMaxPvs( 4 );
   activeMode = 0;
+  buttonIsDown = 0;
 
   controlIsBit = readIsBit = 0;
   prevControlBit = prevReadBit = 0;
@@ -578,6 +579,7 @@ activeGraphicClass *bto = (activeGraphicClass *) this;
   strncpy( minVisString, source->minVisString, 39 );
   strncpy( maxVisString, source->maxVisString, 39 );
   activeMode = 0;
+  buttonIsDown = 0;
 
   prevControlBit = prevReadBit = 0;
   initControlBit = initReadBit = 0;
@@ -2436,6 +2438,8 @@ unsigned int uival;
 
   if ( !controlPvId->have_write_access() ) return;
 
+  buttonIsDown = 0;
+
   if ( toggle ) return;
 
   value = 0;
@@ -2484,6 +2488,8 @@ unsigned int uival;
 
   if ( buttonNumber != 1 ) return;
 
+  buttonIsDown = 1;
+
   if ( controlExists && controlIsBit ) {
 
     curControlBit = ( ( controlV & ( 1 << controlBitPos ) ) > 0 );
@@ -3037,6 +3043,10 @@ int activeButtonClass::setProperty (
 
 char *activeButtonClass::firstDragName ( void ) {
 
+  // If the button is down and we use the middle-click to drag the PV name, the toggle does not seem to reset.
+  // Ignore the drag event in case the button is down
+  if(buttonIsDown) return NULL;
+
   if ( !enabled ) return NULL;
 
   dragIndex = 0;
@@ -3046,6 +3056,10 @@ char *activeButtonClass::firstDragName ( void ) {
 
 char *activeButtonClass::nextDragName ( void ) {
 
+  // If the button is down and we use the middle-click to drag the PV name, the toggle does not seem to reset.
+  // Ignore the drag event in case the button is down
+  if(buttonIsDown) return NULL;
+
   if ( !enabled ) return NULL;
 
   if ( dragIndex < (int) ( sizeof(dragName) / sizeof(char *) ) - 1 ) {
diff --git a/baselib/button.h b/baselib/button.h
index db345a1..2ace521 100644
--- a/baselib/button.h
+++ b/baselib/button.h
@@ -304,6 +304,7 @@ int controlBitPos, readBitPos; // 0-31
 int prevControlBit, prevReadBit;
 int controlBit, readBit;
 int initControlBit, initReadBit;
+int buttonIsDown; // boolean that indicates if we have received the buttonDown event and not yet received the corresponding buttonUp event.
 
 public:
 
-- 
1.8.1.4

